### PR TITLE
Ticket4913 muon begin end commands

### PR DESCRIPTION
**Pull request by JamesKingWork**
THIS MUST BE REVIEWED BY SOMEONE FROM COMPUTING

### Description of work

Added the relevant pre_cmd functions to hook into begin and end commands to check run information and titles and allow the user to changed them where required.

### To test

https://github.com/ISISComputingGroup/IBEX/issues/4913

### Acceptance Criteria

- [ ] On using g.begin() where I have loaded the relevant muon instrument script, I am prompted for the run title
- [ ] On using g.end() where I have loaded the relevant muon instrument script, I am asked to confirm that the run title is correct
- [ ] IBEX provides the same level of functionality as the muon-specific BEGIN and END commands in OpenGenie (Please see emu_routines.gcl on the isis experimental control shares in muon migration)
- [ ] The IBEX genie_python and instrument script documentation are updated to document the muon-specific behaviour.

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Are there unit tests in place? Are the unit tests small and test the a class in isolation?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?

### Functional Tests

- [ ] Do changes function as described? Add comments below that describe the tests performed.
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has developer documentation been updated if required?

### CODE REVIEW

File: technique/muon/run_control.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
+from __future__ import print_function
+from genie_python import genie as g
+from six.moves import input
+from time import sleep
+
+
+def get_input(prompt):
+    """
+    Get input from the user (allows us to mock it for testing)
```
Comment (Tom-Willemsen) [inline technique/muon/run_control.py]: Can we not patch `input` directly? (if not, add justification in the comment)

File: technique/muon/run_control.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    beamline_pars: dict
+        The sample parameters that is to contain the geometry parameter.
+    """
+    if geometry_par in beamline_pars:
+        old_geometry = beamline_pars[geometry_par]
+    else:
+        old_geometry = ""
+    new_geometry = get_input("Geometry (L or T) {}?".format(old_geometry))
+    if new_geometry != "":
+        g.change_beamline_par(geometry_par, new_geometry)
```
Comment (Tom-Willemsen) [inline technique/muon/run_control.py]: Should this check that the input is actually one of L or T and retry if not?

File: technique/muon/run_control.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        if run_title.lower() == "n":
+            g.change_title(get_input("Run title: "))
+        while True:
+            print("Run information:\n")
+            show_label()
+            label_correct = get_input("Is the run information correct (y/n)?")
+            if label_correct.lower() == "y":
+                break
+            else:
+                print("SETTING LABEL")
```
Comment (Tom-Willemsen) [inline technique/muon/run_control.py]: Is this a leftover debug statement?

File: technique/muon/run_control.py (hunk 1 above)
Comment (JamesKingWork) [reply technique/muon/run_control.py]: We cannot patch with `test_WHEN_end_pre_cmd_run_information_incorrect_twice_THEN_set_label_called_twice` as it requires a more complex side effect than patching will allow us to do. Added to comment.

File: technique/muon/run_control.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    Parameters
+    ----------
+    quiet: bool
+      If true suppress run title question output
+    """
+    if not quiet:
+        run_title = get_input("Run title: ")
+        g.change_title(run_title)
+        set_label()
+    sleep(1)
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply technique/muon/run_control.py]: Sleep was called in the equivalent OpenGenie command and before I was not sure why. But with another look, I believe it was waiting for a ClearSeciGraphs procedure to fully finish. Removing now.

File: technique/muon/run_control.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+        while True:
+            print("Run information:\n")
+            show_label()
+            label_correct = get_input("Is the run information correct (y/n)?")
+            if label_correct.lower() == "y":
+                break
+            else:
+                print("SETTING LABEL")
+                set_label()
+    sleep(1)
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply technique/muon/run_control.py]: Sleep was called in the equivalent OpenGenie command and before I was not sure why. But with another look, I believe it was waiting for a ClearSeciGraphs procedure to fully finish. Removing now.

File: technique/muon/run_control.py (hunk 2 above)
Comment (JamesKingWork) [reply technique/muon/run_control.py]: Yes, added.

File: technique/muon/run_control.py (hunk 3 above)
Comment (JamesKingWork) [reply technique/muon/run_control.py]: Yes, removed

File: technique/muon/run_control.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+      The user input
+    """
+    return input(prompt)
+
+# The keys for parameters in the sample parameters dictionary
+name_par = "name"
+orient_par = "geometry"
+temp_par = "temp_label"
+field_par = "field_label"
+comment_par = "comments"
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply technique/muon/run_control.py]: Added to sample params in https://github.com/ISISComputingGroup/EPICS-isisdae/pull/62

File: technique/muon/run_control.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+    str
+      The user input
+    """
+    return input(prompt)
+
+# The keys for parameters in the sample parameters dictionary
+name_par = "name"
+orient_par = "geometry"
+temp_par = "temp_label"
+field_par = "field_label"
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply technique/muon/run_control.py]: Added sample param in https://github.com/ISISComputingGroup/EPICS-isisdae/pull/62

File: technique/muon/run_control.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+    -------
+    str
+      The user input
+    """
+    return input(prompt)
+
+# The keys for parameters in the sample parameters dictionary
+name_par = "name"
+orient_par = "geometry"
+temp_par = "temp_label"
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply technique/muon/run_control.py]: Added sample param in https://github.com/ISISComputingGroup/EPICS-isisdae/pull/62

File: technique/muon/run_control.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+
+
+def set_rb_num():
+    """
+    Ask the user if the rb number is correct and allow them to change it if not.
+    """
+    old_rb_num = g.get_rb()
+    new_rb_num = get_input("RBNo {}?".format(old_rb_num))
+    if new_rb_num != "":
+        g.change_rb(new_rb_num)
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply technique/muon/run_control.py]: Agreed. Fixed.

File: technique/muon/run_control.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+
+    Parameters
+    ----------
+    sample_pars: dict
+        The sample parameters that is to contain the name parameter.
+    """
+    if name_par in sample_pars:
+        old_sample_name = sample_pars[name_par]
+    else:
+        old_sample_name = ""
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply technique/muon/run_control.py]: Much cleaner, did not realise there was a method for this, but it certainly makes sense that there is.

File: technique/muon/tests/test_run_control.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+        run_control.set_temp_sample_par(old_sample_pars)
+
+        # Assert
+        self.assertFalse(g.change_sample_par.called, "Should not have called to set sample param")
+
+    @patch.object(g, "change_title")
+    @patch('technique.muon.run_control.get_input', return_value="n")
+    def test_WHEN_end_pre_cmd_not_quiet_label_correct_THEN_input_called_once(self, _, __):
+        # Act
+        run_control.end_precmd(quiet=False)
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply technique/muon/tests/test_run_control.py]: Thanks for spotting, I have now refactored to make sure I am calling correctly and testing the correct part.

File: technique/muon/tests/test_run_control.py [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+                return "y"
+        return "n"
+
+    @patch.object(run_control, "set_label")
+    def test_WHEN_end_pre_cmd_run_information_incorrect_twice_THEN_set_label_called_twice(self, set_label):
+        # Arrange
+        run_control.get_input = MagicMock(side_effect=self.end_pre_cmd_get_input)
+
+        # Act
+        run_control.end_precmd(quiet=False)
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply technique/muon/tests/test_run_control.py]: Same as above

File: technique/muon/muon_begin_end.py [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+
+    Parameters
+    ----------
+    sample_pars: dict
+        The sample parameters that is to contain the comments parameter.
+    """
+    if comment_par in sample_pars:
+        old_comments = sample_pars[comment_par]
+    else:
+        old_comments = ""
```
Comment (DominicOram) [inline technique/muon/muon_begin_end.py]: As above can use `get`

File: technique/muon/tests/test_muon_begin_end.py [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+    @patch.object(muon_begin_end, "set_label")
+    def test_WHEN_end_pre_cmd_run_information_incorrect_twice_THEN_set_label_called_twice(self, set_label):
+        # Arrange
+        muon_begin_end.get_input = MagicMock(side_effect=self.end_pre_cmd_get_input)
+
+        # Act
+        muon_begin_end.end_precmd()
+
+        # Assert
+        self.assertEquals(set_label.call_count, 2, "Run information rejected twice so should have asked "
```
Comment (DominicOram) [inline technique/muon/tests/test_muon_begin_end.py]:
```suggestion
        self.assertEqual(set_label.call_count, 2, "Run information rejected twice so should have asked "
```
`self.assertEquals` is depreciated. Instead use `self.assertEqual`.

File: technique/muon/muon_begin_end.py (hunk 13 above)
Comment (JamesKingWork) [reply technique/muon/muon_begin_end.py]: Thank you for picking that up, missed it!

**Comment by JamesKingWork** [comment]
> All works well. **Very** minor changes which I would do myself but not sure I could push them to your repo?

All fixed :)

**Pull request merged**